mempool/src/pool/mod.rs

```diff
@@ -263,77 +272,78 @@ where
-            return Err(TxValidationError::ExceedsMaxBlockSize);
-        }
+        let chainstate_handle =
+            subsystem::blocking::BlockingHandle::new(self.chainstate_handle().shallow_clone());
 
-        if self.contains_transaction(&tx.transaction().get_id()) {
-            return Err(TxValidationError::TransactionAlreadyInMempool);
-        }
+        let tip = chainstate_handle.call(|c| c.get_best_block_id())??;
 
-        let tx = TxWithFee::new(self, tx.clone()).await?;
-        let conflicts = self.rbf_checks(&tx)?;
+        let current_best = chainstate_handle
+            .call(move |c| c.get_gen_block_index(&tip))??
+            .ok_or(TxValidationError::InternalError)?;
 
-        self.verify_inputs_available(tx.tx()).await?;
+        let mut tx_verifier = self.tx_verifier.derive_child();
 
-        self.pays_minimum_relay_fees(&tx)?;
+        let fee = tx_verifier
+            .connect_transaction(
+                &TransactionSourceForConnect::Mempool {
+                    current_best: &current_best,
+                },
+                &tx,
+                &BlockTimestamp::from_duration_since_epoch(self.clock.get_time()),
+            )?
+            .ok_or(TxValidationError::FeeNotDetermined)?;
 
+        let tx = TxWithFee::new_with_fee(tx, fee.into());
+        self.pays_minimum_relay_fees(&tx)?;
         self.pays_minimum_mempool_fee(&tx)?;
 
-        Ok(conflicts)
-    }
+        let conflicts = if ENABLE_RBF {
+            self.rbf_checks(&tx)?
+        } else {
+            // Without RBF enabled, any conflicting transaction results in an error
+            ensure!(
+                self.conflicting_tx_ids(tx.tx()).next().is_none(),
+                TxValidationError::ConflictWithIrreplaceableTransaction
+            );
+            Conflicts::new(BTreeSet::new())
+        };
 
-    async fn verify_inputs_available(
-        &self,
-        tx: &SignedTransaction,
-    ) -> Result<(), TxValidationError> {
-        let tx_clone = tx.clone();
-        let chainstate_inputs = self
-            .chainstate_handle
-            .call(move |this| this.available_inputs(tx_clone.transaction()))
-            .await??;
-        tx.transaction()
-            .inputs()
-            .iter()
-            .find(|input| {
-                !chainstate_inputs.contains(&Some((*input).clone()))
-                    && !self.store.contains_outpoint(input.outpoint())
-            })
-            .map_or_else(
-                || Ok(()),
-                |input| {
-                    Err(TxValidationError::OutPointNotFound {
-                        outpoint: input.outpoint().clone(),
-                        spending_tx_id: tx.transaction().get_id(),
-                    })
-                },
-            )
+        let delta = tx_verifier.consume()?;
+
+        let final_tip = chainstate_handle.call(|c| c.get_best_block_id())??;
+        ensure!(tip == final_tip, TxValidationError::TipMoved);
```

Comment:
I don't like the way the new tip is processed here, because valid transactions may be rejected unexpectedly. Isn't it better to try again instead?

Reply:
I see two possibilities here:

1. Retry adding the transaction if tip moved, with a limited number of attempts.
2. Change the tx verifier in mempool to pre-fetch all data needed to validate a transaction in one `call` to get a consistent view.

In the long run, we should probably go for (2) but I think (1) will do in the meantime.

Reply:
> I see two possibilities here:
> 
> 1. Retry adding the transaction if tip moved, with a limited number of attempts.
> 2. Change the tx verifier in mempool to pre-fetch all data needed to validate a transaction in one `call` to get a consistent view.
> 
> In the long run, we should probably go for (2) but I think (1) will do in the meantime.

Implemented approach (1) for the time of being